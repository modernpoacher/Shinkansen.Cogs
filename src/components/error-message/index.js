/**
 * ErrorMessage component
 */
import React, { Component } from 'react'
import PropTypes from 'prop-types'

import TextContent from 'shinkansen-cogs/components/common/text-content'

export default class ErrorMessage extends Component {
  hasTextContent () {
    const { errorMessage } = this.props

    return !!errorMessage
  }

  getTextContent () {
    const { errorMessage } = this.props

    return errorMessage
  }

  getClassName () {
    return 'error-message'
  }

  shouldComponentUpdate (props) {
    return (
      (props.errorMessage !== this.props.errorMessage)
    )
  }

  renderTextContent () {
    if (this.hasTextContent()) {
      const textContent = this.getTextContent()

      return (
        <TextContent textContent={textContent} />
      )
    }

    return null
  }

  render () {
    const { errorMessage } = this.props

    if (errorMessage) {
      return (
        <span className={this.getClassName()}>
          {this.renderTextContent()}
        </span>
      )
    }

    return null
  }
}

ErrorMessage.propTypes = {
  errorMessage: PropTypes.string
}
