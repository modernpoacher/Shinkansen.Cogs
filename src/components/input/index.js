/**
 * Input component
 */
import React, { Component } from 'react'
import PropTypes from 'prop-types'

export default class Input extends Component {
  getDOMNode = () => this.domNode
  setDOMNode = (domNode) => !!(this.domNode = domNode) || delete this.domNode

  getClassName () {
    return 'input'
  }

  shouldComponentUpdate (props) {
    return (
      (props.id !== this.props.id) ||
      (props.required !== this.props.required) ||
      (props.disabled !== this.props.disabled) ||
      (props.readOnly !== this.props.readOnly) ||
      (props.tabIndex !== this.props.tabIndex) ||
      (props.accessKey !== this.props.accessKey) ||
      (props.placeholder !== this.props.placeholder) ||
      (props.onChange !== this.props.onChange)
    )
  }

  render () {
    const {
      id,
      required,
      disabled,
      readOnly,
      tabIndex,
      accessKey,
      placeholder
    } = this.props

    return (
      <input
        id={id}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        placeholder={placeholder}
        className={this.getClassName()}
        ref={this.setDOMNode}
      />
    )
  }
}

Input.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  label: PropTypes.string,
  required: PropTypes.bool,
  disabled: PropTypes.bool,
  readOnly: PropTypes.bool,
  tabIndex: PropTypes.number,
  accessKey: PropTypes.string,
  placeholder: PropTypes.string,
  onChange: PropTypes.func
}

Input.defaultProps = {
  label: 'Input',
  required: false,
  disabled: false,
  readOnly: false,
  onChange: () => {}
}

/**
 * CheckInput component
 */
export class CheckInput extends Input {
  getValue () {
    return this.getDOMNode()
      .checked
  }

  setValue (checked) {
    this.getDOMNode()
      .checked = checked
  }

  shouldComponentUpdate (props) {
    return (
      super.shouldComponentUpdate(props) ||
      (props.checked !== this.props.checked) ||
      (props.onClick !== this.props.onClick)
    )
  }

  handleClick = ({ target: { checked } }) => {
    const { onClick } = this.props

    onClick(checked)
  }

  handleChange = ({ target: { checked } }) => {
    const { onChange } = this.props

    onChange(checked)
  }
}

CheckInput.propTypes = {
  ...Input.propTypes,
  checked: PropTypes.bool,
  defaultChecked: PropTypes.bool,
  onClick: PropTypes.func
}

CheckInput.defaultProps = {
  ...Input.defaultProps
}

/**
 * ValueInput component
 */
export class ValueInput extends Input {
  getValue () {
    return this.getInput()
      .value
  }

  setValue (value) {
    this.getInput()
      .value = value
  }

  shouldComponentUpdate (props) {
    return (
      super.shouldComponentUpdate(props) ||
      (props.value !== this.props.value)
    )
  }

  handleChange = ({ target: { value } }) => {
    const { onChange } = this.props

    onChange(value)
  }
}

ValueInput.propTypes = {
  ...Input.propTypes,
  value: PropTypes.string,
  defaultValue: PropTypes.string
}

ValueInput.defaultProps = {
  ...Input.defaultProps
}
