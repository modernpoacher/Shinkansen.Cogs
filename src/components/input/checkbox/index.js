/**
 * CheckboxInput component
 */
import React from 'react'
import classnames from 'classnames'
import { CheckInput } from 'shinkansen-cogs/components/input'

export default class CheckboxInput extends CheckInput {
  getClassName () {
    return classnames(super.getClassName(), 'checkbox')
  }

  render () {
    const {
      id,
      name,
      checked,
      defaultChecked,
      required,
      disabled,
      readOnly,
      tabIndex,
      accessKey
    } = this.props

    const className = this.getClassName()

    return (
      <input
        id={id}
        name={name}
        checked={checked}
        defaultChecked={defaultChecked}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        onClick={this.handleClick}
        onChange={this.handleChange}
        className={className}
        type='checkbox'
        ref={this.setDOMNode}
      />
    )
  }
}

CheckboxInput.propTypes = {
  ...CheckInput.propTypes
}

CheckboxInput.defaultProps = {
  ...CheckInput.defaultProps,
  label: 'Checkbox'
}
