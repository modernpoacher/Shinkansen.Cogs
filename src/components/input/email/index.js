/**
 * EmailInput component
 */
import React from 'react'
import classnames from 'classnames'
import { ValueInput } from 'shinkansen-cogs/components/input'

export default class EmailInput extends ValueInput {
  getClassName () {
    return classnames(super.getClassName(), 'email')
  }

  render () {
    const {
      id,
      name,
      value,
      defaultValue,
      required,
      disabled,
      readOnly,
      tabIndex,
      accessKey,
      placeholder
    } = this.props

    const className = this.getClassName()

    return (
      <input
        id={id}
        name={name}
        value={value}
        defaultValue={defaultValue}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        placeholder={placeholder}
        onChange={this.handleChange}
        className={className}
        type='email'
        ref={this.setInput}
      />
    )
  }
}

EmailInput.propTypes = {
  ...ValueInput.propTypes
}

EmailInput.defaultProps = {
  ...ValueInput.defaultProps
}
