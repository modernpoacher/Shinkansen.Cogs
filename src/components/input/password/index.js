/**
 * PasswordInput component
 */
import React from 'react'
import classnames from 'classnames'
import { ValueInput } from 'shinkansen-cogs/components/input'

export default class PasswordInput extends ValueInput {
  getClassName () {
    return classnames(super.getClassName(), 'password')
  }

  render () {
    const {
      id,
      name,
      value,
      defaultValue,
      required,
      disabled,
      readOnly,
      tabIndex,
      accessKey,
      placeholder
    } = this.props

    const className = this.getClassName()

    return (
      <input
        id={id}
        name={name}
        value={value}
        defaultValue={defaultValue}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        placeholder={placeholder}
        onChange={this.handleChange}
        className={className}
        type='password'
        ref={this.setInput}
      />
    )
  }
}

PasswordInput.propTypes = {
  ...ValueInput.propTypes
}

PasswordInput.defaultProps = {
  ...ValueInput.defaultProps,
  label: 'Password'
}
