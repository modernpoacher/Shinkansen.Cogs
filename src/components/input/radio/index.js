/**
 * RadioInput component
 */
import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { CheckInput } from 'shinkansen-cogs/components/input'

export default class RadioInput extends CheckInput {
  getClassName () {
    return classnames(super.getClassName(), 'radio')
  }

  handleClick = ({ target: { value } }) => {
    const { onClick } = this.props

    onClick(value)
  }

  handleChange = ({ target: { value } }) => {
    const { onChange } = this.props

    onChange(value)
  }

  render () {
    const {
      id,
      name,
      value,
      checked,
      defaultChecked,
      required,
      disabled,
      readOnly,
      tabIndex,
      accessKey
    } = this.props

    const className = this.getClassName()

    return (
      <input
        id={id}
        name={name}
        value={value}
        checked={checked}
        defaultChecked={defaultChecked}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        onClick={this.handleClick}
        onChange={this.handleChange}
        className={className}
        type='radio'
        ref={this.setDOMNode}
      />
    )
  }
}

RadioInput.propTypes = {
  ...CheckInput.propTypes,
  id: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired
}

RadioInput.defaultProps = {
  ...CheckInput.defaultProps
}
