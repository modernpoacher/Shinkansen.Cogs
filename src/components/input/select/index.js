/**
 * SelectInput component
 */
import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { ValueInput } from 'shinkansen-cogs/components/input'

export default class SelectInput extends ValueInput {
  getClassName () {
    return classnames(super.getClassName(), 'select')
  }

  render () {
    const {
      id,
      name,
      value,
      defaultValue,
      required,
      disabled,
      readOnly,
      tabIndex,
      accessKey,
      children
    } = this.props

    const className = this.getClassName()

    return (
      <select
        id={id}
        name={name}
        value={value}
        defaultValue={defaultValue}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        onChange={this.handleChange}
        className={className}
        ref={this.setInput}>
        {children}
      </select>
    )
  }
}

SelectInput.propTypes = {
  ...ValueInput.propTypes,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(
      PropTypes.element
    )
  ])
}

SelectInput.defaultProps = {
  ...ValueInput.defaultProps
}
