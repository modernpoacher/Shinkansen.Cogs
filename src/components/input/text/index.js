/**
 * TextInput component
 */
import React from 'react'
import classnames from 'classnames'
import { ValueInput } from 'shinkansen-cogs/components/input'

export default class TextInput extends ValueInput {
  getClassName () {
    return classnames(super.getClassName(), 'text')
  }

  render () {
    const {
      id,
      name,
      value,
      defaultValue,
      required,
      disabled,
      readOnly,
      placeholder
    } = this.props

    const className = this.getClassName()

    return (
      <input
        id={id}
        name={name}
        value={value}
        defaultValue={defaultValue}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        placeholder={placeholder}
        onChange={this.handleChange}
        className={className}
        type='text'
        ref={this.setInput}
      />
    )
  }
}

TextInput.propTypes = {
  ...ValueInput.propTypes
}

TextInput.defaultProps = {
  ...ValueInput.defaultProps,
  label: 'Text'
}
