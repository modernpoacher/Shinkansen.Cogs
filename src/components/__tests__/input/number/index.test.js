import React, { Component as mockComponent } from 'react'
import renderer from 'react-test-renderer'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import classnames from 'classnames'

import { ValueInput } from 'shinkansen-cogs/components/input'
import Input from 'shinkansen-cogs/components/input/number'

Enzyme.configure({ adapter: new Adapter() })

jest.mock('classnames', () => jest.fn(() => 'MOCK CLASSNAME'))

jest.mock('shinkansen-cogs/components/input', () => {
  class MockInput extends mockComponent {
    getClassName () { }
  }

  return {
    __esModule: true,
    ValueInput: class ValueInput extends MockInput { },
    default: MockInput
  }
})

describe('shinkansen-cogs/components/input/number', () => {
  describe('<Input />', () => {
    describe('With required props', () => {
      const component = (
        <Input name='MOCK NAME' />
      )

      it('renders', () => {
        expect(renderer.create(component).toJSON())
          .toMatchSnapshot()
      })

      describe('`getClassName`', () => {
        it('is defined', () => {
          expect(Input.prototype.getClassName)
            .toBeDefined()
        })
      })
    })

    describe('With additional props', () => {
      it('renders', () => {
        const component = (
          <Input
            id='MOCK ID'
            name='MOCK NAME'
            label='MOCK LABEL'
            required
            disabled
            readOnly
          />
        )

        expect(renderer.create(component).toJSON())
          .toMatchSnapshot()
      })
    })

    describe('`getClassName()`', () => {
      let returnValue

      beforeEach(() => {
        jest.spyOn(ValueInput.prototype, 'getClassName').mockReturnValue('MOCK GETCLASSNAME')

        const component = (
          <Input name='MOCK NAME' />
        )

        const instance = (
          shallow(component)
            .instance()
        )

        returnValue = instance.getClassName()
      })

      it('invokes `classnames`', () => {
        expect(classnames)
          .toBeCalledWith('MOCK GETCLASSNAME', 'number')
      })

      it('returns the classname', () => {
        expect(returnValue)
          .toBe('MOCK CLASSNAME')
      })
    })
  })
})
