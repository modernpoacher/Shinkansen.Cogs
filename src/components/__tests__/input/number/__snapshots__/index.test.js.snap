// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shinkansen-cogs/components/input/number <Input /> With additional props renders 1`] = `
<input
  className="MOCK CLASSNAME"
  disabled={true}
  id="MOCK ID"
  name="MOCK NAME"
  onChange={[Function]}
  readOnly={true}
  required={true}
  type="number"
/>
`;

exports[`shinkansen-cogs/components/input/number <Input /> With required props renders 1`] = `
<input
  className="MOCK CLASSNAME"
  disabled={false}
  name="MOCK NAME"
  onChange={[Function]}
  readOnly={false}
  required={false}
  type="number"
/>
`;
